### ISSUE TITLE
Runtime differences with different backends

### ISSUE THREAD

**Issue by kwaters4** [pnnl/lamellar-runtime#27]
Here is the code that the problem was reproduced with, but I think it has to do with the backends.

```Rust
use lamellar::array::prelude::*;


fn main() {
    let world = lamellar::LamellarWorldBuilder::new().build();
    let my_pe = world.my_pe();
    let array_size: usize = 5000;
    let arr = AtomicArray::<usize>::new(&world, array_size, Distribution::Block);

    arr.block_on(arr.store(my_pe, my_pe + 1));
    if my_pe == 0 {
        println!("starting collect");
        let v: Vec<usize> = arr.onesided_iter().into_iter().map (|&x| x).collect();
        if v[my_pe] == my_pe + 1 {
            println!("passed check on pe {}", my_pe);
        }
        else {
            println!("failed check on pe {}", my_pe);
        }
    }
    arr.wait_all();
    println!("done on pe {}", my_pe);
}
```

Lamellar Version
``` 
dev branch
0.50
```
Rust version 
``` rustc 1.68.1 (8460ca823 2023-03-20)```

### Local
cargo run: 0.1s 

### Shared memory
with -N=1 -T=10: 0.4s
with -N=2 -T=5: 27s
with -N=2 -T=10: 17s
with -N=10 -T=1: 2:08m

### ROFI
-N 1 --mpi=pmi2 : 2:30m
-N 2 --mpi=pmi2 5:15m
-n 10 : 8:34m
-n 20 : 9:50m

**Comment by rdfriese** [comment]
While I expect there to be some variation between backends, with each level of communication (local -> shared memory -> distributed) introducing slowdowns, the range you are showing is definitely surprising. 

(*disclaimer* sorry for the following wall of text, its mostly to help me reason about debugging, but may be generally useful to others as well)

There could be a few things causing this:

- didn't compile with `--release`  -- while this can cause pretty significant performance improvement I would be shocked if it caused such variability in runtime
- An obvious to me (but unfortunately probably not to anyone else, as the docs don't really do much to point it out) is that `onesided_iter()` is a _very_ communication heavy operation as every single iteration will trigger a single element transfer, after the change from 1 PE to multiple PEs, though I actually expect that the performance should be relatively constant.  In most cases it is recommend to use a [buffered_onesided_iter](https://docs.rs/lamellar/latest/lamellar/array/iterator/trait.LamellarArrayIterators.html#tymethod.buffered_onesided_iter). My gut feeling is that this is not the main issue though.
- For at least the Rofi backend, initialization currently has to perform essentially a serialized handshake when registering the PEs with one another, resulting in linear scaling as number of PEs increase (but this should only be a one time cost when calling `.build()` ), at the scales show above, though this shouldn't even be on the order of 10s of seconds, so I doubt this is the main issue
- Executing on a numa architecture and not binding individual threads to cores, or at least numa domains -- I have seen this cause pretty big fluctuations in performance, but it typically shows up when executing a large  number of active messages, the scale of this example simply seems too small to really see significant effects
- oversubscribing more than one Lamellar Thread per physical core (allocated to the PE process), of the issues listed, this is my best guess,  there can be significant performance implications if the number of Lamellar Threads exceeds the number of physical cores available to the process (the range you are seeing still seems very severe).  I think there may be an off by one error between the environment variable "LAMELLAR_THREADS" and the number of threads being launched (i.e. I dont think we consider the master thread), a simple hack would be to set LAMELLAR_THREADS equal to one less than the number of cores available to that PE.  For example if I have a 20 Core Node, and are using 2 PEs, each PE would get 10 Cores, so I would set `LAMELLAR_THREADS=9`. 

TL;DR
make sure to compile with `--release` 
set `LAMELLAR_THREADS ="number of cores (per pe) - 1"`

If you are able to pull dev again, I've updated the `lamellar_run.sh` script to fix a few issues, and automatically perform the "-1" for the LAMELLAR_THREADS environment variable. I'm currently looking to see if my hypothesis on the off by one is the underlying issue.

**Comment by rdfriese** [comment]
I've created a new branch associated with this issue that should better handle oversubscribing threads to cores (assuming this was the actual underlying issue). For what its worth when I performed the same tests on my end I saw variation in the runtimes as well, but not to the same extent as yours. With these new changes, the runtime spread is much tighter for me. I will be interested if it is the same for you.
